File: driver-management-api/app/api/v1/safety.py
from fastapi import APIRouter, Depends, HTTPException, Query, BackgroundTasks
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List, Optional
from datetime import datetime, timedelta
from app.core.database import get_db
from app.core.deps import get_current_user
from app.core.logging import logger
from app.models import User
from app.schemas import (
    GPSRecordCreate, GPSRecordResponse,
    DrivingBehaviorCreate, DrivingBehaviorUpdate, DrivingBehaviorResponse,
    EmergencyAlertCreate, EmergencyAlertUpdate, EmergencyAlertResponse
)

router = APIRouter()


@router.get("/stats")
async def get_safety_stats(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """获取安全统计（仪表板期望字段）"""
    logger.info(f"用户 {current_user.username} 请求安全统计")
    return {
        "online_drivers": 0,
        "active_vehicles": 0,
        "today_alerts": 0,
        "emergency_alerts": 0
    }


@router.get("/alerts/recent")
async def get_recent_alerts(
    limit: int = Query(10, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """获取最近预警 - 简化版本，返回 data 包装"""
    logger.info(f"用户 {current_user.username} 请求最近预警，限制: {limit}")
    return {"data": []}


@router.post("/gps-records", response_model=GPSRecordResponse)
async def create_gps_record(
    gps_record: GPSRecordCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """创建GPS记录 - 简化版本"""
    logger.info(f"用户 {current_user.username} 尝试创建GPS记录")
    return GPSRecordResponse(
        id=1,
        vehicle_id=gps_record.vehicle_id,
        driver_id=gps_record.driver_id,
        latitude=gps_record.latitude,
        longitude=gps_record.longitude,
        speed=gps_record.speed,
        direction=gps_record.direction,
        altitude=gps_record.altitude,
        timestamp=gps_record.timestamp or datetime.now(),
        created_at=datetime.now()
    )


@router.get("/gps-records", response_model=List[GPSRecordResponse])
async def get_gps_records(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    vehicle_id: Optional[int] = Query(None),
    driver_id: Optional[int] = Query(None),
    start_time: Optional[datetime] = Query(None),
    end_time: Optional[datetime] = Query(None),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """获取GPS记录列表 - 简化版本"""
    logger.info(f"用户 {current_user.username} 请求GPS记录列表")
    return []


@router.post("/driving-behaviors", response_model=DrivingBehaviorResponse)
async def create_driving_behavior(
    behavior: DrivingBehaviorCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """创建驾驶行为记录 - 简化版本"""
    logger.info(f"用户 {current_user.username} 尝试创建驾驶行为记录")
    return DrivingBehaviorResponse(
        id=1,
        driver_id=behavior.driver_id,
        vehicle_id=behavior.vehicle_id,
        behavior_type=behavior.behavior_type,
        severity=behavior.severity,
        description=behavior.description,
        location=behavior.location,
        latitude=behavior.latitude,
        longitude=behavior.longitude,
        recorded_at=behavior.recorded_at or datetime.now(),
        created_at=datetime.now()
    )


@router.get("/driving-behaviors", response_model=List[DrivingBehaviorResponse])
async def get_driving_behaviors(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    driver_id: Optional[int] = Query(None),
    vehicle_id: Optional[int] = Query(None),
    behavior_type: Optional[str] = Query(None),
    severity: Optional[str] = Query(None, pattern="^(low|medium|high|critical)$"),
    start_date: Optional[datetime] = Query(None),
    end_date: Optional[datetime] = Query(None),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """获取驾驶行为记录列表 - 简化版本"""
    logger.info(f"用户 {current_user.username} 请求驾驶行为记录列表")
    return []


@router.post("/emergency-alerts", response_model=EmergencyAlertResponse)
async def create_emergency_alert(
    alert: EmergencyAlertCreate,
    background_tasks: BackgroundTasks,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """创建紧急预警 - 简化版本"""
    logger.info(f"用户 {current_user.username} 尝试创建紧急预警")
    return EmergencyAlertResponse(
        id=1,
        alert_type=alert.alert_type,
        severity=alert.severity,
        title=alert.title,
        description=alert.description,
        driver_id=alert.driver_id,
        vehicle_id=alert.vehicle_id,
        location=alert.location,
        latitude=alert.latitude,
        longitude=alert.longitude,
        status="active",
        created_at=datetime.now(),
        updated_at=datetime.now()
    )


@router.get("/emergency-alerts", response_model=List[EmergencyAlertResponse])
async def get_emergency_alerts(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    status: Optional[str] = Query(None, pattern="^(active|resolved|ignored)$"),
    severity: Optional[str] = Query(None, pattern="^(low|medium|high|critical)$"),
    driver_id: Optional[int] = Query(None),
    vehicle_id: Optional[int] = Query(None),
    start_date: Optional[datetime] = Query(None),
    end_date: Optional[datetime] = Query(None),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """获取紧急预警列表 - 简化版本"""
    logger.info(f"用户 {current_user.username} 请求紧急预警列表")
    return []


@router.put("/emergency-alerts/{alert_id}", response_model=EmergencyAlertResponse)
async def update_emergency_alert(
    alert_id: int,
    alert_update: EmergencyAlertUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """更新紧急预警 - 简化版本"""
    logger.info(f"用户 {current_user.username} 尝试更新紧急预警 {alert_id}")
    return EmergencyAlertResponse(
        id=alert_id,
        alert_type="sos",
        severity="high",
        title="测试预警",
        description="这是一个测试预警",
        status=alert_update.status or "resolved",
        created_at=datetime.now(),
        updated_at=datetime.now()
    )

File: driver-management-vue/src/api/safety.ts
import api from './auth'
import type { 
  GPSRecordResponse, GPSRecordCreate,
  DrivingBehaviorResponse, DrivingBehaviorCreate, DrivingBehaviorUpdate,
  EmergencyAlertResponse, EmergencyAlertCreate, EmergencyAlertUpdate 
} from '@/types/safety'

// 使用统一的axios实例，自动携带认证信息

// GPS轨迹管理相关API
export async function createGPSRecord(gpsRecord: GPSRecordCreate): Promise<GPSRecordResponse> {
  const response = await api.post(`/safety/gps-records`, gpsRecord)
  return response.data
}

export async function getGPSRecords(
  skip = 0,
  limit = 100,
  vehicleId?: number,
  driverId?: number,
  startTime?: Date,
  endTime?: Date
): Promise<GPSRecordResponse[]> {
  const params = new URLSearchParams()
  params.append('skip', skip.toString())
  params.append('limit', limit.toString())
  if (vehicleId) params.append('vehicle_id', vehicleId.toString())
  if (driverId) params.append('driver_id', driverId.toString())
  if (startTime) params.append('start_time', startTime.toISOString())
  if (endTime) params.append('end_time', endTime.toISOString())
  
  const response = await api.get(`/safety/gps-records`, { params })
  return response.data
}

export async function getVehicleTrack(
  vehicleId: number,
  startTime: Date,
  endTime: Date
): Promise<{
  vehicle_id: number
  plate_number: string
  start_time: Date
  end_time: Date
  total_points: number
  total_distance: number
  max_speed: number
  avg_speed: number
  idle_time: number
  track: Array<{
    latitude: number
    longitude: number
    speed: number
    timestamp: Date
    address: string
    status: string
  }>
}> {
  const params = new URLSearchParams()
  params.append('start_time', startTime.toISOString())
  params.append('end_time', endTime.toISOString())
  
  const response = await api.get(`/safety/vehicles/${vehicleId}/track`, { params })
  return response.data
}

// 驾驶行为管理相关API
export async function createDrivingBehavior(behavior: DrivingBehaviorCreate): Promise<DrivingBehaviorResponse> {
  const response = await api.post(`/safety/driving-behaviors`, behavior)
  return response.data
}

export async function getDrivingBehaviors(
  skip = 0,
  limit = 100,
  driverId?: number,
  vehicleId?: number,
  behaviorType?: 'harsh_braking' | 'speeding' | 'sharp_turn',
  severity?: 'low' | 'medium' | 'high',
  processed?: boolean,
  startTime?: Date,
  endTime?: Date
): Promise<DrivingBehaviorResponse[]> {
  const params = new URLSearchParams()
  params.append('skip', skip.toString())
  params.append('limit', limit.toString())
  if (driverId) params.append('driver_id', driverId.toString())
  if (vehicleId) params.append('vehicle_id', vehicleId.toString())
  if (behaviorType) params.append('behavior_type', behaviorType)
  if (severity) params.append('severity', severity)
  if (processed !== undefined) params.append('processed', processed.toString())
  if (startTime) params.append('start_time', startTime.toISOString())
  if (endTime) params.append('end_time', endTime.toISOString())
  
  const response = await api.get(`/safety/driving-behaviors`, { params })
  return response.data
}

export async function updateDrivingBehavior(
  behaviorId: number,
  behaviorUpdate: DrivingBehaviorUpdate
): Promise<DrivingBehaviorResponse> {
  const response = await api.put(`/safety/driving-behaviors/${behaviorId}`, behaviorUpdate)
  return response.data
}

export async function getDrivingBehaviorSummary(
  driverId?: number,
  vehicleId?: number,
  startDate?: Date,
  endDate?: Date
): Promise<{
  total_behaviors: number
  by_type: Record<string, number>
  by_severity: Record<string, number>
  processed_count: number
  unprocessed_count: number
  high_risk_percentage: number
}> {
  const params = new URLSearchParams()
  if (driverId) params.append('driver_id', driverId.toString())
  if (vehicleId) params.append('vehicle_id', vehicleId.toString())
  if (startDate) params.append('start_date', startDate.toISOString())
  if (endDate) params.append('end_date', endDate.toISOString())
  
  const response = await api.get(`/safety/driving-behaviors/summary`, { params })
  return response.data
}

// 紧急警报管理相关API
export async function createEmergencyAlert(alert: EmergencyAlertCreate): Promise<EmergencyAlertResponse> {
  const response = await api.post(`/safety/emergency-alerts`, alert)
  return response.data
}

export async function getEmergencyAlerts(
  skip = 0,
  limit = 100,
  driverId?: number,
  vehicleId?: number,
  alertType?: 'accident' | 'medical' | 'security',
  severity?: 'low' | 'medium' | 'high' | 'critical',
  status?: 'active' | 'responded' | 'resolved'
): Promise<EmergencyAlertResponse[]> {
  const params = new URLSearchParams()
  params.append('skip', skip.toString())
  params.append('limit', limit.toString())
  if (driverId) params.append('driver_id', driverId.toString())
  if (vehicleId) params.append('vehicle_id', vehicleId.toString())
  if (alertType) params.append('alert_type', alertType)
  if (severity) params.append('severity', severity)
  if (status) params.append('status', status)
  
  const response = await api.get(`/safety/emergency-alerts`, { params })
  return response.data
}

export async function updateEmergencyAlert(
  alertId: number,
  alertUpdate: EmergencyAlertUpdate
): Promise<EmergencyAlertResponse> {
  const response = await api.put(`/safety/emergency-alerts/${alertId}`, alertUpdate)
  return response.data
}

export async function getActiveEmergencyAlertsSummary(): Promise<{
  total_active_alerts: number
  by_type: Record<string, number>
  by_severity: Record<string, number>
  latest_alerts: Array<{
    id: number
    driver_name: string
    alert_type: string
    severity: string
    description: string
    created_at: Date
    latitude: number
    longitude: number
  }>
}> {
  const response = await api.get(`/safety/emergency-alerts/active-summary`)
  return response.data
}

// 新增用于仪表板的API函数
export async function getSafetyStats(): Promise<{
  online_drivers: number
  active_vehicles: number
  today_alerts: number
  emergency_alerts: number
}> {
  const response = await api.get(`/safety/stats`)
  return response.data
}

export async function getRecentAlerts(params?: {
  limit?: number
  severity?: string
  status?: string
}): Promise<{
  data: Array<{
    id: number
    alert_type: string
    severity: string
    status: string
    description: string
    created_at: string
    driver?: {
      id: number
      name: string
    }
  }>
  total: number
}> {
  const queryParams = new URLSearchParams()
  if (params?.limit) queryParams.append('limit', params.limit.toString())
  if (params?.severity) queryParams.append('severity', params.severity)
  if (params?.status) queryParams.append('status', params.status)
  
  const response = await api.get(`/safety/alerts/recent?${queryParams}`)
  return response.data
}

export async function getAlerts(params?: {
  page?: number
  page_size?: number
  severity?: string
  status?: string
  driver_id?: number
}): Promise<{
  data: Array<{
    id: number
    alert_type: string
    severity: string
    status: string
    description: string
    created_at: string
    driver?: {
      id: number
      name: string
    }
  }>
  total: number
}> {
  const queryParams = new URLSearchParams()
  if (params?.page) queryParams.append('page', params.page.toString())
  if (params?.page_size) queryParams.append('page_size', params.page_size.toString())
  if (params?.severity) queryParams.append('severity', params.severity)
  if (params?.status) queryParams.append('status', params.status)
  if (params?.driver_id) queryParams.append('driver_id', params.driver_id.toString())
  
  const response = await api.get(`/safety/alerts?${queryParams}`)
  return response.data
}

export async function processAlert(alertId: number): Promise<void> {
  await api.put(`/safety/alerts/${alertId}/process`)
}

export async function getEmergencyStats(): Promise<{
  today_emergency: number
  pending_emergency: number
  resolved_emergency: number
}> {
  const response = await api.get(`/safety/emergency-alerts/stats`)
  return response.data
}

export async function processEmergencyAlert(alertId: number): Promise<void> {
  await api.put(`/safety/emergency-alerts/${alertId}/process`)
}